### ISSUE TITLE
Geo-aware data fetching from replica

### ISSUE THREAD

**Issue by canhld94** [ClickHouse/ClickHouse#52398]
Let say, we a cluster with 3 replicas, one in Asia and two in NA. Is there anyway to force the replicas in NA to only fetch parts from the other replica in NA instead of the Asia one? This will reduce the network cost significantly.

Example: we have 3 replica ASIA1, NA1, NA2. When inserting data to ASIA1, we don't want NA1 and NA2 to concurrently fetch from ASIA1, but NA1 will fetch first from ASIA1, then NA2 will fetch from NA1 only.

I've been searching the issues list but there's no discussion before this issue. One of my idea is similar to hedged read: to fetch a part, we first open connections to all possible replicas, then select the one with easiest response. This can somehow provide a sense of location-aware. But we still need to control the replica flow.

Appreciate for any advices 🙏

**Comment by UnamedRus** [comment]
There is `load_balancing='nearest_hostname'` setting for queries

Probably we can have similar setting to replication parts exchange.
But it will be more complicated, what if both replicas NA1 & NA2 doesn't have that part and schedule fetch from ASIA1 at the same moment. (ie one of NA replicas should wait)

**Comment by canhld94** [comment]
> But it will be more complicated, what if both replicas NA1 & NA2 doesn't have that part and schedule fetch from ASIA1 at the same moment. (ie one of NA replicas should wait)

Yes, this is our headache now. Seem no workaround solution now unless we develop one.

**Comment by canhld94** [comment]
So after discussion, we think our solution may look like this:

```
NA1 (NA regional leader) <------------------> APAC1 (APAC regional leader)
         |                                                |
  NA2 (follower)                                  APAC2 (follower)
```

Each region will have a regional leader. All replicas still pull log entries from the common zk table log queue, but when coming fetching, a regional follower can only fetch from the corresponding region leader.

Now the question is how to do leader election for each region. May be don't need a fancy one, zk-based or even rule-based should satisfy the requirements.
